File: tail-recursion.py
'''
Implement tail recursion instead of trivial recursion
Case in point: factorial of n
'''

import sys


def go(n, a):
    ''' Helper function'''
    return a if (n == 1) else go(n - 1, a * n)


def fact(n):
    ''' factorial function'''
    return go (n, 1)


if __name__ == '__main__':
    li = sys.argv
    if len(li) != 2:
        print(f'Usage: `./1-tail_recursion N` where N is a positive integer\n')
        sys.exit(1)
    else:
        fact(li[1])

File: tail-recursion.c
#include <stdio.h>

/**
*Calculate the factorial of a number using tail recursion instead of trivial recursion
* fact: fn to calc factorial of n
*@n: positive int
*return: int: factorial of n
*go: helper fuction; eliminates the need to have trivial recursion
*@n: positive int
*@a: cumulative product
*/

int fact (int n);
int go (int n, int a);

int main (int __attribute__ ((unused)) argc, int argv){
    int len = 0;
    while (argv) {
        len++;
    }
    
    if (len != 2) {
        printf('Usage: `./0-tail_recursion N` where N is a positive integer\n');
        return 1;
    } else {
        return fact(argv[1]);
    }
}

fact(n){
    return go(n, 1);
}

go(n, a){
    return n == 1 ? a : go(n - 1, a * n);
}